### PR TITLE
Add `pipeline` package.

### LINKED ISSUE #111: Add commands and timeout messages created by handlers to the in-memory queue.

### PR DESCRIPTION
**Pull request by jmalloc**
This PR adds the `pipeline.Source` type which is an abstraction literally for a "source of sessions" that can be implemented on top of the event stream and the message queue.

A source produces sessions represented by the `pipeline.Session` interface.

`queue.Queue.Pop()` and `queue.Session` have been changed to implement `pipeline.Source` and `pipeline.Session`, respectively.

Fixes #111, partially addresses #74.

**Note:** This package was originally called `session`, not `pipeline`, hence the name of the branch.

### CODE REVIEW

File: pipeline/handler_test.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+		}
+
+		stage := Handle(h)
+
+		err := stage(context.Background(), scope)
+		Expect(err).ShouldNot(HaveOccurred())
+		Expect(called).To(BeTrue())
+	})
+
+	It("returns an error if the envelope can not be unpacked", func() {
```
Comment (danilvpetrov) [inline pipeline/handler_test.go:67]:
```suggestion
	It("returns an error if the envelope cannot be unpacked", func() {
```

File: pipeline/pipeline.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
+package pipeline
+
+import (
+	"context"
+)
+
+// A Source provides message sessions.
```
Comment (danilvpetrov) [inline pipeline/pipeline.go:7]:
```suggestion
// Source provides message sessions.
```

File: pipeline/pipeline.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+type Source func(context.Context) (Session, error)
+
+// A Sink accepts message sessions.
```
Comment (danilvpetrov) [inline pipeline/pipeline.go:10]:
```suggestion
// Sink accepts message sessions.
```

File: pipeline/pipeline_test.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+func pass(context.Context, *Scope) error {
+	return nil
+}
+
+// fail is a Sink that always returns an error.
+func fail(context.Context, *Scope) error {
+	return errors.New("<error: fail() called>")
+}
+
+// noop is stage that forwards to the next stage without doing anything.
```
Comment (danilvpetrov) [inline pipeline/pipeline_test.go:22]:
```suggestion
// noop is a stage that forwards to the next stage without doing anything.
```

File: pipeline/pump.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+import (
+	"context"
+
+	"github.com/dogmatiq/dodeca/logging"
+	"github.com/dogmatiq/marshalkit"
+	"golang.org/x/sync/errgroup"
+)
+
+// Pump obtains sessions a source and feeds them to a sink.
```
Comment (danilvpetrov) [inline pipeline/pump.go:11]: Does it obtain a session source or does it obtain a source for sessions?

File: pipeline/session.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+import (
+	"context"
+	"time"
+
+	"github.com/dogmatiq/infix/envelope"
+	"github.com/dogmatiq/infix/persistence"
+)
+
+// A Session is a specialized transaction that encapsulates an atomic
```
Comment (danilvpetrov) [inline pipeline/session.go:11]:
```suggestion
// Session is a specialized transaction that encapsulates an atomic
```

File: pipeline/pipeline.go (hunk 2 above)
Comment (jmalloc) [reply pipeline/pipeline.go:7]: This is intentional and valid in Go.

File: pipeline/pump.go (hunk 5 above)
Comment (jmalloc) [reply pipeline/pump.go:11]: Should say "from a source", thx :)

File: pipeline/pipeline.go (hunk 2 above)
Comment (danilvpetrov) [reply pipeline/pipeline.go:7]: Ok, I did not know that.

File: pipeline/pump.go (hunk 5 above)
Comment (jmalloc) [inline pipeline/pump.go:11]:
```suggestion
// Pump obtains sessions from a source and feeds them to a sink.
```

**Pull request merged**